--- a/libinotifytools/src/inotifytools.c
+++ b/libinotifytools/src/inotifytools.c
@@ -34,6 +34,14 @@
 
 #include "inotifytools/inotify.h"
 
+#ifndef stat64
+#define stat64 stat
+#endif
+
+#ifndef lstat64
+#define lstat64 lstat
+#endif
+
 /**
  * @file inotifytools/inotifytools.h
  * @brief inotifytools library public interface.
--- a/src/common.c
+++ b/src/common.c
@@ -13,6 +13,14 @@
 
 #include <inotifytools/inotifytools.h>
 
+#ifndef stat64
+#define stat64 stat
+#endif
+
+#ifndef lstat64
+#define lstat64 lstat
+#endif
+
 #define MAXLEN 4096
 #define LIST_CHUNK 1024
 
