Index: sedutil-2022-12-27-7a0cda7f/Common/DtaOptions.h
===================================================================
--- sedutil-2022-12-27-7a0cda7f.orig/Common/DtaOptions.h
+++ sedutil-2022-12-27-7a0cda7f/Common/DtaOptions.h
@@ -21,6 +21,10 @@ along with sedutil.  If not, see <http:/
 #ifndef _DTAOPTIONS_H
 #define	_DTAOPTIONS_H
 
+#ifndef uint8_t
+#include <cstdint>
+#endif
+
 /** Output modes */
 typedef enum _sedutiloutput {
 	sedutilNormal,
