Index: tesseract-4.1.1/src/dict/matchdefs.h
===================================================================
--- tesseract-4.1.1.orig/src/dict/matchdefs.h
+++ tesseract-4.1.1/src/dict/matchdefs.h
@@ -24,6 +24,10 @@
 #include <cstdio>
 #include "unichar.h"
 
+#ifndef uint16_t
+#include <cstdint>
+#endif
+
 /* define the maximum number of classes defined for any matcher
   and the maximum class id for any matcher. This must be changed
   if more different classes need to be classified */
